File: app/src/main/java/com/iutcalendar/event/DialogPopupEvent.java
package com.iutcalendar.event;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.util.Log;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.calendar.iutcalendar.R;
import com.iutcalendar.calendrier.EventCalendrier;
import com.iutcalendar.task.ClickListener;
import com.iutcalendar.task.PersonnalCalendrier;
import com.iutcalendar.task.Task;
import com.iutcalendar.task.TaskRecycleView;

public class DialogPopupEvent extends Dialog {
    private final EventCalendrier relatedEvent;
    private final Activity activity;
    private TextView title, summary, salle, horaire, duree;
    private Button okBtn;
    private ImageButton addBtn;
    private RecyclerView recyclerViewTask;
    private final UpdateListener whenFinish;

    public DialogPopupEvent(@NonNull Context context, EventCalendrier eventClicked, Activity activity, UpdateListener whenFinish) {
        super(context);
        Log.d("Dialog", "start");
        setContentView(R.layout.dialog_event_edit);
        initVariable();
        this.relatedEvent = eventClicked;
        this.activity = activity;
        this.whenFinish = whenFinish;

        this.addBtn.setOnClickListener(view -> showAddTask());

        this.okBtn.setOnClickListener(view -> dismiss());


        String timeDebut = relatedEvent.getDate().timeToString();
        String timeFin = relatedEvent.getDate().addTime(relatedEvent.getDuree()).timeToString();

        this.title.setText(relatedEvent.getSummary());
        this.summary.setText(relatedEvent.getDescription());
        this.salle.setText(relatedEvent.getSalle().replace("\\,", ", "));

        this.horaire.setText(context.getString(R.string.both_time, timeDebut, timeFin));
        this.duree.setText(relatedEvent.getDuree().timeToString());

        updatedTask();
        Log.d("Dialog", "end");
    }

    private void updatedTask() {
        ClickListener listener = this::removeTask;
        TaskRecycleView adapter = new TaskRecycleView(PersonnalCalendrier.getInstance(getContext()).getLinkedTask(relatedEvent), listener);
        recyclerViewTask.setAdapter(adapter);
        recyclerViewTask.setLayoutManager(new LinearLayoutManager(activity));
        Log.d("Dialog", "updateTask");
    }

    @Override
    public void dismiss() {
        whenFinish.update();
        super.dismiss();
    }

    private void initVariable() {
        title = findViewById(R.id.title);
        summary = findViewById(R.id.textSummary);
        salle = findViewById(R.id.salle);
        horaire = findViewById(R.id.horaire);
        duree = findViewById(R.id.duree);


        okBtn = findViewById(R.id.okBtn);
        addBtn = findViewById(R.id.addBtn);

        recyclerViewTask = findViewById(R.id.listTask);
    }

    private void showAddTask() {
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(getContext());
        alertDialog.setTitle("Ajouté un tâche");
        alertDialog.setMessage(getContext().getString(R.string.task));

        final EditText editText = new EditText(getContext());
        LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT);
        editText.setLayoutParams(lp);
        alertDialog.setView(editText);

        alertDialog.setPositiveButton(getContext().getString(R.string.submit),
                (dialog, which) -> {
                    PersonnalCalendrier.getInstance(getContext()).addLinkedTask(new Task(editText.getText().toString(), relatedEvent.getUID()), relatedEvent);
                    PersonnalCalendrier.getInstance(getContext()).save(getContext());
                    dialog.dismiss();
                    updatedTask();
                });

        alertDialog.setNegativeButton(getContext().getString(R.string.cancel),
                (dialog, which) -> dialog.cancel());

        alertDialog.show();
    }

    private void removeTask(Task taskClicked) {
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(getContext());
        if (!taskClicked.isAlarm()) {
            alertDialog.setTitle("Suppression");
            alertDialog.setMessage("Voulez-vous supprimer cette tâche ?\n\"" + taskClicked.getTxt() + "\"");

            alertDialog.setPositiveButton(getContext().getString(R.string.yes), (dialogInterface, i) -> {

                PersonnalCalendrier.getInstance(getContext()).remove(getContext(), taskClicked);
                PersonnalCalendrier.getInstance(getContext()).save(getContext());
                Toast.makeText(getContext(), "Tâche supprimer", Toast.LENGTH_SHORT).show();
                updatedTask();
                dialogInterface.dismiss();
            });
        } else {
            alertDialog.setTitle("Désactivation");
            alertDialog.setMessage("Voulez-vous " + ((taskClicked.isAlarmActivate()) ? "désactivé" : "activé") +
                    " cette tâche ?\n\"" + taskClicked.getTxt() + "\"");

            alertDialog.setPositiveButton(getContext().getString(R.string.yes), (dialogInterface, i) -> {

                taskClicked.setAlarmActivate(!taskClicked.isAlarmActivate());
                PersonnalCalendrier.getInstance(getContext()).save(getContext());
                updatedTask();
                dialogInterface.dismiss();

            });
        }


        alertDialog.setNegativeButton(getContext().getString(R.string.no), (dialogInterface, i) -> dialogInterface.cancel());

        alertDialog.show();
    }


}

File: app/src/main/java/com/iutcalendar/event/DialogPopupEvent.kt
package com.iutcalendar.event

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.drawable.ColorDrawable
import android.util.Log
import android.view.View
import android.widget.*
import androidx.core.graphics.blue
import androidx.core.graphics.green
import androidx.core.graphics.red
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.iutcalendar.calendrier.EventCalendrier
import com.iutcalendar.data.CachedData
import com.iutcalendar.data.ColorEvent
import com.iutcalendar.dialog.ColorPickerDialog
import com.iutcalendar.task.PersonalCalendrier
import com.iutcalendar.task.Task
import com.iutcalendar.task.TaskAdapter
import com.univlyon1.tools.agenda.R

class DialogPopupEvent(
    context: Context,
    eventClicked: EventCalendrier,
    activity: Activity?,
    private val whenFinish: (shouldRestart: Boolean) -> Unit
) : Dialog(context) {
    private val relatedEvent: EventCalendrier
    private val activity: Activity?
    private lateinit var title: TextView
    private lateinit var summary: TextView
    private lateinit var salle: TextView
    private lateinit var horaire: TextView
    private lateinit var duree: TextView
    private lateinit var tpsRestant: TextView
    private lateinit var okBtn: Button
    private lateinit var addBtn: ImageButton
    private lateinit var recyclerViewTask: RecyclerView
    private lateinit var colorEdit: View
    private var shouldRestart: Boolean = false


    init {
        Log.d("Dialog", "start")
        setContentView(R.layout.dialog_event_edit)
        initVariable()
        relatedEvent = eventClicked
        this.activity = activity
        addBtn.setOnClickListener { showAddTask() }
        okBtn.setOnClickListener { dismiss() }
        val timeDebut = relatedEvent.date!!.timeToString()
        val timeFin = relatedEvent.date!!.addTime(relatedEvent.dure).timeToString()
        title.text = relatedEvent.nameEvent
        summary.text = relatedEvent.description
        salle.text = relatedEvent.salle.replace("\\,", ", ")
        horaire.text = context.getString(R.string.both_time, timeDebut, timeFin)
        duree.text = relatedEvent.dure.timeToString()

        countMinutesModule().let {
            tpsRestant.text = "${it / 60}h ${it % 60}min"
        }


        updateColorEdit()
        colorEdit.setOnClickListener {
            (colorEdit.background as ColorDrawable).color.apply {
                ColorPickerDialog(context, red, green, blue) { colorId ->
                    ColorEvent.save(context, relatedEvent.nameEvent, colorId)
                    updateColorEdit()
                    shouldRestart = true
                }.show()
            }
        }



        initRecyclerViewTask()
        Log.d("Dialog", "end")
    }

    private fun initVariable() {
        title = findViewById(R.id.title)
        summary = findViewById(R.id.textSummary)
        salle = findViewById(R.id.salle)
        horaire = findViewById(R.id.horaire)
        duree = findViewById(R.id.duree)
        tpsRestant = findViewById(R.id.tpsRestant)
        okBtn = findViewById(R.id.okBtn)
        addBtn = findViewById(R.id.addBtn)
        recyclerViewTask = findViewById(R.id.listTask)
        colorEdit = findViewById(R.id.colorEdit)
    }

    private fun countMinutesModule(): Int {
        var count = 0
        var startCount = false
        CachedData.calendrier.events.forEach { e ->
            if (e.uid == relatedEvent.uid) startCount = true
            if (startCount && e.nameEvent == relatedEvent.nameEvent) count += e.dure.hour * 60 + e.dure.minute
        }
        return count
    }

    private fun initRecyclerViewTask() {
        val taskAdapter = TaskAdapter(context, relatedEvent)
        recyclerViewTask.apply {
            adapter = taskAdapter
            layoutManager = LinearLayoutManager(activity)
        }
        Log.d("Dialog", "updateTask")
    }

    private fun updateColorEdit() {
        colorEdit.setBackgroundColor(ColorEvent.getOrCreate(context, relatedEvent.nameEvent))//should get and not create
    }

    override fun dismiss() {
        whenFinish(shouldRestart)
        super.dismiss()
    }

    private fun showAddTask() {
        val alertDialog = AlertDialog.Builder(context)
        alertDialog.setTitle("Ajouté un tâche")
        alertDialog.setMessage(context.getString(R.string.task))
        val editText = EditText(context)
        editText.layoutParams = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.MATCH_PARENT
        )
        alertDialog.setView(editText)
        alertDialog.setPositiveButton(
            context.getString(R.string.submit)
        ) { dialog: DialogInterface, _: Int ->
            PersonalCalendrier.getInstance(context).apply {
                addLinkedTask(Task(editText.text.toString(), relatedEvent.uid), relatedEvent)
                save(context)
            }
            dialog.dismiss()
            recyclerViewTask.adapter?.apply {
                notifyItemInserted(itemCount - 1)
            }
        }
        alertDialog.setNegativeButton(context.getString(R.string.cancel)) { dialog: DialogInterface, _: Int -> dialog.cancel() }
        alertDialog.show()
    }

    private fun changeColor() {


    }

}